api/apps/api/src/modules/scenarios/scenarios.controller.ts

```diff
@@ -308,23 +270,10 @@ export class ScenariosController {
       authenticatedUser: req.user,
     });
     if (isLeft(result)) {
-      switch (result.left) {
-        case projectNotReady:
-          throw new ConflictException();
-        case projectDoesntExist:
-          throw new NotFoundException(`Project doesn't exist`);
-        case forbiddenError:
-          throw new ForbiddenException(
-            `User with ID: ${req.user.id} is not allowed to create scenarios`,
-          );
-        case blmCreationFailure:
-          throw new InternalServerErrorException(
-            `Could not create initial BLM for scenario`,
-          );
-
-        default:
-          throw new InternalServerErrorException();
-      }
+      return aclErrorHandler(result.left, {
```

Comment:
formally, `aclErrorHandler()`'s return type should be `Never`, so the `return` here (and in all the occurrences in this controller) should be redundant and slightly misleading to the reader of this piece of code.

another opinionated opinion here is that although `aclErrorHandler()` tidies up a lot of scattered error handling in one single neatly organized place, we end up throwing out of sight. I'm looking at this on GH so I'm likely to suggest some nonsense, but what about instead using `aclErrorHandler()` as an error translation layer (so it could be `mapAclDomainErrorToHttpError()`, maybe?), where each branch of the `switch` statement would *return* a `new HttpErrorOfSomeKind(options)`, which we would then `throw` in the controller functions?

so here specifically, this would be

```
throw mapAclDomainErrorToHttpError(result.left, {
  userId: req.user.id,
  resourceType: 'scenarios',
});
```

I am not super comfortable with this either, and I haven't really ever seen anything like this in JS or other languages, so I may just be hallucinating. The idea of a domain-to-http error translation layer sounds appealing to me in terms of readability/maintainability, but it may be just me 😬 

Also, `resourceType` may be set from `scenarioResource.name.plural`

Reply:
well, since I had that nagging feeling that I was reinventing something from first principles, my brain went on a further neural flare and I eventually _remembered_ that NestJS gives us the lovely _exception filters_ functionality.

All is good and let's merge - but, maybe, maybe, the use of expressive errors via `Maybe`/`Either` could be pushed to the very end by throwing a generic `AclError` (or something like this) everywhere in these controllers, passing to it whatever `left` values we have there, and move the mapping from domain errors to HTTP exceptions to an exception filter.

Absolutely not _now_ of course, but I have the feeling this would be the ultimate logical evolution of the `Either` setup. Something to chew on 😄